tokio/src/runtime/time/entry.rs

```diff
@@ -485,13 +494,39 @@ impl TimerEntry {
     fn inner(&self) -> &TimerShared {
         let inner = unsafe { &*self.inner.get() };
         if inner.is_none() {
+            let shard_id = self.get_shard_id();
             unsafe {
-                *self.inner.get() = Some(TimerShared::new());
+                *self.inner.get() = Some(TimerShared::new(shard_id));
             }
         }
         return inner.as_ref().unwrap();
     }
 
+    // Gets the shard id. If current thread is a worker thread, we use its worker index as a shard id.
+    // Otherwise, we use a random number generator to obtain the shard id.
+    cfg_rt! {
+        fn get_shard_id(&self) -> u32 {
+            let shard_size = self.driver.driver().time().inner.get_shard_size();
+            let id = context::with_scheduler(|ctx| match ctx {
+                Some(scheduler::Context::CurrentThread(_ctx)) => 0,
+                #[cfg(feature = "rt-multi-thread")]
+                Some(scheduler::Context::MultiThread(ctx)) => ctx.get_worker_index() as u32,
+                #[cfg(all(tokio_unstable, feature = "rt-multi-thread"))]
+                Some(scheduler::Context::MultiThreadAlt(ctx)) => ctx.get_worker_index() as u32,
+                _ => context::thread_rng_n(shard_size),
+            });
+            id % shard_size
+        }
+    }
```

Comment:
This sounds like a getter for a value that is already created, but actually it is the logic for figuring out which shard to use for a new timer. Can we rename this, and perhaps move it to a stand-alone function instead of a method on `TimerEntry`. Otherwise I am worried that someone will call it and expect it to return the same value as what the `TimerShared` is using.